File: sdks/go/streaming.go
package dmrx

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

// Stream provides channel-based access to streaming response chunks.
type Stream struct {
	ch     chan StreamChunk
	errCh  chan error
	doneCh chan struct{}
}

// Ch returns a read-only channel of stream chunks.
func (s *Stream) Ch() <-chan StreamChunk {
	return s.ch
}

// Err returns a read-only channel for stream errors.
func (s *Stream) Err() <-chan error {
	return s.errCh
}

// Done returns a channel that closes when the stream is complete.
func (s *Stream) Done() <-chan struct{} {
	return s.doneCh
}

// Iter returns a channel for range-based iteration over stream chunks.
func (s *Stream) Iter() <-chan StreamChunk {
	return s.ch
}

// parseSSELine parses a single SSE line into a key-value pair.
func parseSSELine(line string) (string, string, bool) {
	line = strings.TrimSpace(line)
	if line == "" {
		return "", "", false
	}
	if idx := strings.Index(line, ":"); idx >= 0 {
		key := strings.TrimSpace(line[:idx])
		value := strings.TrimSpace(line[idx+1:])
		return key, value, true
	}
	return "", "", false
}

// parseStreamEvent parses an SSE data payload into a StreamChunk.
func parseStreamEvent(dataStr string) (*StreamChunk, error) {
	// Handle OpenAI end marker
	if dataStr == "[DONE]" {
		return &StreamChunk{
			Type:  StreamChunkDone,
			Data:  nil,
			Index: 0,
		}, nil
	}

	var raw map[string]interface{}
	if err := json.Unmarshal([]byte(dataStr), &raw); err != nil {
		return nil, fmt.Errorf("failed to parse stream event: %w", err)
	}

	// DMR-X native streaming format
	if chunkType, ok := raw["type"].(string); ok {
		chunkData, _ := json.Marshal(raw["data"])
		index := 0
		if idx, ok := raw["index"].(float64); ok {
			index = int(idx)
		}

		return &StreamChunk{
			Type:  StreamChunkType(chunkType),
			Data:  chunkData,
			Index: index,
		}, nil
	}

	// OpenAI-compatible streaming format
	if obj, ok := raw["object"].(string); ok && obj == "chat.completion.chunk" {
		choices, _ := raw["choices"].([]interface{})
		if len(choices) > 0 {
			choice, _ := choices[0].(map[string]interface{})

			// Check for finish reason (stream done)
			if finishReason, ok := choice["finish_reason"].(string); ok && finishReason != "" {
				doneData := DoneData{
					RequestID:    getString(raw, "id"),
					ModelID:      getString(raw, "model"),
					FinishReason: finishReason,
				}
				dataBytes, _ := json.Marshal(doneData)
				return &StreamChunk{
					Type:  StreamChunkDone,
					Data:  dataBytes,
					Index: 0,
				}, nil
			}

			// Token delta
			delta, _ := choice["delta"].(map[string]interface{})
			tokenData := TokenData{}
			if content, ok := delta["content"].(string); ok {
				tokenData.Content = &content
			}
			if tc, ok := delta["tool_calls"].([]interface{}); ok {
				tcBytes, _ := json.Marshal(tc)
				json.Unmarshal(tcBytes, &tokenData.ToolCalls)
			}
			if role, ok := delta["role"].(string); ok {
				tokenData.Role = role
			}

			dataBytes, _ := json.Marshal(tokenData)
			return &StreamChunk{
				Type:  StreamChunkToken,
				Data:  dataBytes,
				Index: 0,
			}, nil
		}
	}

	return nil, fmt.Errorf("unknown stream format: %s", dataStr)
}

// readStream reads SSE lines and sends parsed chunks to the channel.
func readStream(reader io.Reader, stream *Stream) {
	defer close(stream.doneCh)
	defer close(stream.ch)

	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	var dataBuffer string

	for scanner.Scan() {
		line := scanner.Text()

		// Skip empty lines (SSE message separators)
		if line == "" {
			if dataBuffer != "" {
				chunk, err := parseStreamEvent(dataBuffer)
				if err != nil {
					stream.errCh <- err
					return
				}
				if chunk.Type == StreamChunkDone && chunk.Data == nil {
					// [DONE] marker
					return
				}
				stream.ch <- *chunk
				dataBuffer = ""
			}
			continue
		}

		// Parse SSE field
		key, value, ok := parseSSELine(line)
		if !ok {
			continue
		}

		switch key {
		case "event":
			// Skip event type lines
		case "data":
			dataBuffer = value
		case ":":
			// SSE comment, skip
		}
	}

	// Handle trailing data
	if dataBuffer != "" {
		chunk, err := parseStreamEvent(dataBuffer)
		if err == nil {
			if !(chunk.Type == StreamChunkDone && chunk.Data == nil) {
				stream.ch <- *chunk
			}
		}
	}

	if err := scanner.Err(); err != nil {
		select {
		case stream.errCh <- fmt.Errorf("stream scanner error: %w", err):
		default:
		}
	}
}

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

func getFloat(m map[string]interface{}, key string) float64 {
	if v, ok := m[key].(float64); ok {
		return v
	}
	return 0
}

File: sdks/python/dmrx/streaming.py
"""
Streaming response parsing for DMR-X API.
Handles SSE (Server-Sent Events) and line-delimited JSON streams.
"""

from __future__ import annotations

import json
import time
from typing import Any, AsyncIterator, Dict, Iterator, Optional

from .callbacks import CallbackManager
from .errors import DMRXError, map_error
from .types.shared import (
    DoneStreamChunk,
    ErrorStreamChunk,
    StreamChunk,
    TokenStreamChunk,
)


# ── SSE Parsing ────────────────────────────────────────────────────


def parse_sse_line(line: str) -> Optional[Dict[str, Any]]:
    """Parse a single SSE line into a key-value pair."""
    line = line.strip()
    if not line:
        return None
    colon_idx = line.find(":")
    if colon_idx == -1:
        return {line: ""}
    key = line[:colon_idx].strip()
    value = line[colon_idx + 1 :].strip()
    return {key: value}


def parse_sse_event(data_str: str) -> Optional[StreamChunk]:
    """
    Parse an SSE data payload into a structured StreamChunk.

    Supports:
    - OpenAI format: data: {"choices": [...]}
    - DMR-X native format: data: {"type": "token", "data": ..., "index": ...}
    - Anthropic format: event: content_block_delta / data: {...}
    """

    # DMR-X native format (chunks have 'type' and 'data' at top level)
    try:
        parsed = json.loads(data_str)
    except json.JSONDecodeError:
        return None

    if not isinstance(parsed, dict):
        return None

    chunk_type = parsed.get("type")

    # DMR-X native streaming chunks
    if chunk_type in ("token", "done", "error", "image_partial", "video_partial", "audio_chunk", "audio_partial"):
        if chunk_type == "token":
            return TokenStreamChunk(
                type="token",
                data=parsed.get("data", {}),
                index=parsed.get("index", 0),
            )
        elif chunk_type == "done":
            return DoneStreamChunk(
                type="done",
                data=parsed.get("data", {}),
                index=parsed.get("index", 0),
            )
        elif chunk_type == "error":
            return ErrorStreamChunk(
                type="error",
                data=parsed.get("data", {}),
                index=parsed.get("index", 0),
            )
        else:
            return StreamChunk(
                type=chunk_type,  # type: ignore
                data=parsed.get("data", {}),
                index=parsed.get("index", 0),
            )

    # OpenAI-compatible streaming format
    # data: {"id":"...","object":"chat.completion.chunk","choices":[{"delta":{...}}]}
    if parsed.get("object") == "chat.completion.chunk":
        choices = parsed.get("choices", [])
        content = ""
        tool_calls = []

        for choice in choices:
            delta = choice.get("delta", {})
            delta_content = delta.get("content")
            if delta_content is not None:
                content += delta_content
            delta_tool_calls = delta.get("tool_calls")
            if delta_tool_calls:
                tool_calls.extend(delta_tool_calls)

        finish_reason = choices[0].get("finish_reason") if choices else None

        if finish_reason:
            return DoneStreamChunk(
                type="done",
                data={
                    "requestId": parsed.get("id", ""),
                    "modelId": parsed.get("model", ""),
                    "finishReason": finish_reason,
                },
                index=0,
            )

        return TokenStreamChunk(
            type="token",
            data={
                "content": content if content else None,
                "tool_calls": tool_calls if tool_calls else None,
            },
            index=0,
        )

    return None


# ── Sync Stream ────────────────────────────────────────────────────


class Stream:
    """
    Synchronous streaming response iterator.

    Usage:
        with client.chat.stream(ChatCompletionRequest(...)) as stream:
            for chunk in stream:
                if isinstance(chunk, TokenStreamChunk):
                    print(chunk.data.get("content", ""), end="")
    """

    def __init__(
        self,
        lines: Iterator[str],
        *,
        endpoint: str = "",
        request_body: Optional[Dict[str, Any]] = None,
        callbacks: Optional[CallbackManager] = None,
    ) -> None:
        self._lines = lines
        self._endpoint = endpoint
        self._request_body = request_body or {}
        self._callbacks = callbacks or CallbackManager()
        self._done = False
        self._start_time = time.monotonic()

    def __iter__(self) -> Iterator[StreamChunk]:
        return self._iter_chunks()

    def _iter_chunks(self) -> Iterator[StreamChunk]:
        buffer = ""

        for raw_line in self._lines:
            line = raw_line.strip()

            # Skip empty lines
            if not line:
                continue

            # Handle SSE event fields
            if line.startswith("event:"):
                # Anthropic-style events (event: ...)
                continue

            if line.startswith("data:"):
                data_str = line[5:].strip()

                # OpenAI stream end marker
                if data_str == "[DONE]":
                    self._done = True
                    duration = (time.monotonic() - self._start_time) * 1000
                    self._callbacks.on_stream_end(
                        self._endpoint,
                        self._request_body,
                        {"finishReason": "stop"},
                        duration,
                    )
                    return

                chunk = parse_sse_event(data_str)
                if chunk is not None:
                    yield from self._emit_chunk(chunk)
            elif line.startswith(":"):
                # SSE comment, skip
                continue
            else:
                # Plain JSON line (non-SSE)
                chunk = parse_sse_event(line)
                if chunk is not None:
                    yield from self._emit_chunk(chunk)

    def _emit_chunk(self, chunk: StreamChunk) -> Iterator[StreamChunk]:
        """Yield a chunk, firing callbacks as appropriate."""
        if isinstance(chunk, ErrorStreamChunk):
            duration = (time.monotonic() - self._start_time) * 1000
            error = DMRXError(
                chunk.data.get("message", "Stream error"),
                code=chunk.data.get("code", "STREAM_ERROR"),
            )
            self._callbacks.on_stream_error(
                self._endpoint, self._request_body, error, duration
            )
            yield chunk
            return

        if isinstance(chunk, DoneStreamChunk):
            self._done = True
            duration = (time.monotonic() - self._start_time) * 1000
            self._callbacks.on_stream_end(
                self._endpoint, self._request_body, chunk.data, duration
            )
            yield chunk
            return

        # Token or other intermediate chunk
        self._callbacks.on_stream_chunk(
            self._endpoint, self._request_body, chunk.model_dump()
        )
        yield chunk

    def close(self) -> None:
        """Close the underlying iterator if possible."""
        if hasattr(self._lines, "close"):
            self._lines.close()


# ── Async Stream ───────────────────────────────────────────────────


class AsyncStream:
    """
    Asynchronous streaming response iterator.

    Usage:
        async with client.chat.stream(ChatCompletionRequest(...)) as stream:
            async for chunk in stream:
                if isinstance(chunk, TokenStreamChunk):
                    print(chunk.data.get("content", ""), end="")
    """

    def __init__(
        self,
        lines: AsyncIterator[str],
        *,
        endpoint: str = "",
        request_body: Optional[Dict[str, Any]] = None,
        callbacks: Optional[CallbackManager] = None,
    ) -> None:
        self._lines = lines
        self._endpoint = endpoint
        self._request_body = request_body or {}
        self._callbacks = callbacks or CallbackManager()
        self._done = False
        self._start_time = time.monotonic()

    def __aiter__(self) -> AsyncIterator[StreamChunk]:
        return self._iter_chunks()

    async def _iter_chunks(self) -> AsyncIterator[StreamChunk]:
        async for raw_line in self._lines:
            line = raw_line.strip()

            if not line:
                continue

            if line.startswith("event:"):
                continue

            if line.startswith("data:"):
                data_str = line[5:].strip()

                if data_str == "[DONE]":
                    self._done = True
                    duration = (time.monotonic() - self._start_time) * 1000
                    self._callbacks.on_stream_end(
                        self._endpoint,
                        self._request_body,
                        {"finishReason": "stop"},
                        duration,
                    )
                    return

                chunk = parse_sse_event(data_str)
                if chunk is not None:
                    async for c in self._emit_chunk_async(chunk):
                        yield c
            elif line.startswith(":"):
                continue
            else:
                chunk = parse_sse_event(line)
                if chunk is not None:
                    async for c in self._emit_chunk_async(chunk):
                        yield c

    async def _emit_chunk_async(
        self, chunk: StreamChunk
    ) -> AsyncIterator[StreamChunk]:
        """Yield a chunk asynchronously, firing callbacks as appropriate."""
        if isinstance(chunk, ErrorStreamChunk):
            duration = (time.monotonic() - self._start_time) * 1000
            error = DMRXError(
                chunk.data.get("message", "Stream error"),
                code=chunk.data.get("code", "STREAM_ERROR"),
            )
            self._callbacks.on_stream_error(
                self._endpoint, self._request_body, error, duration
            )
            yield chunk
            return

        if isinstance(chunk, DoneStreamChunk):
            self._done = True
            duration = (time.monotonic() - self._start_time) * 1000
            self._callbacks.on_stream_end(
                self._endpoint, self._request_body, chunk.data, duration
            )
            yield chunk
            return

        # Token or other intermediate chunk
        self._callbacks.on_stream_chunk(
            self._endpoint, self._request_body, chunk.model_dump()
        )
        yield chunk

    async def close(self) -> None:
        """Close the underlying async iterator."""
        if hasattr(self._lines, "aclose"):
            await self._lines.aclose()
